### ISSUE TITLE
Redesign questions

### ISSUE THREAD

**Issue by grimmySwe** [ingydotnet/git-subrepo#142]
@ingydotnet  When I started to write some documentation and tried to explain how git-subrepo works I came to some insights. (I have discussed this a lot with @jrosdahl )

The main reason for these thoughts are that the most common use case where you repeatedly pull things and have some minor local change would be increasingly more difficult. Every time you pull you would find the common ancestor further back in history and you would have to solve the same conflicts over and over again. We already discussed in other issues that it would be nice if you could avoid having an extra branch if you want local changes.
- When performing a pull, it should be possible to use the pull reference in .gitrepo instead of the merge-base command. And instead of rebasing onto the new branch you should be able to simply apply the diff between the current state and the rebase parent. As the commits get squashed anyway.
  - Replace subrepo merge-base with simple use of .gitrepo.commit
  - Replace rebase with merge
  - Simplify subrepo branch

```
D(a) - E(a) - F(a) project   (commit(.gitrepo.commit))
a - b - c subrepo
$ git subrepo pull
* fetch
a - b - c subrepo/fetch
* branch
d - e - f subrepo/branch
* reparent f on a (see further down for reparenting tool)
a - b - c subrepo/fetch
  \
    f subrepo/branch
* merge, solve conflicts
a - b - c - g subrepo/fetch
 \         /
     f            subrepo/branch
* return to project
* write contents from g into subrepodir
* Update .gitrepo file
* commit
D(a) - E(a) - F(a) - G(g) project
```

Note that you if you really want to use rebase instead of merge you can do the reparenting on e instead and rebase the entire branch step by step. It will though add a little complexity as you would need to find out the first commit with the last .gitrepo.commit. I guess that one way would be to iterate from the HEAD and find the first one.
- For the push case it's a little bit trickier, in our use case it might be better to squash the push commit as well. Then you have the freedom of simply using the same simple logic of comparing the changes from the latest pull point and apply them to subrepo. If you keep the condition that you need to pull before you push it might even be simpler as all you need to do is create a new commit with the content of project/subrepo
  - Squash subrepo upstream 
  - Simplify subrepo branch

```
D(a) - E(a) - F(a) project   (commit(.gitrepo.commit))
a subrepo
* fetch
a subrepo/fetch
* branch
d - e - f subrepo/branch
* checkout subrepo/fetch
* write contents of f 
* commit (it will though get a new commit hash)
a - f subrepo/fetch
* push
a - f subrepo
```
- If you really want to push all commits into the subrepo there are some strange things occuring. The commit sequence that appears in the subrepo is not linear, as the changes will occur in a rebased order including possible "pull" commits that reintroduces the changes already existing in the subrepo.

I found a tool for reparenting:
https://github.com/MarkLodato/git-reparent

When I think about it I guess that you could make the branch command work on only the last commit. That way you have constant speed. Current design will increase the time for each commit.

```
$ git checkout -b subrepo/branch
$ git filter-branch --subdirectory-filter subrepo -- HEAD^..HEAD
```

I haven't done any coding yet, all of this came up when I tried to describe how git-subrepo works.

**Comment by grimmySwe** [comment]
And yes I realize that much of this is returning to 0.2.3... solving the problems found then in another way.

Regarding the squash/no-squash it might be goo to have that optional on a subrepo basis. From my personal perspective I think that the squash method would be preferable in a corporate environment where you end up with huge amount of commits. But I do understand that in smaller repositories it might be beneficial to keep everything.

**Comment by grimmySwe** [comment]
One more thing, today there is a -u flag to subrepo push allowing you to update .gitrepo. I think that it might be a good idea to make this default behaviour. And in the case of squashing the push you get the same message on both sides.
```
$ git subrepo push -m "Adding feature X" bar
```
If you don't pass -m you could open up git_editor with the commit listing. This commit listing might need a previous push reference.

That way you could work by only pushing if no one else is working in the repo.

**Comment by grimmySwe** [comment]
more thinking...
If you decide to push and want all commits visible in subrepo, it might be useful to use the rebase --interactive style, so you can actually remove uneccessary commits/pulls.

**Comment by grimmySwe** [comment]
Tested some parent stuff and this seems to be working
```
$ git checkout master
$ git checkout -b subrepo/shared/branch
$ git filter-branch --subdirectory-filter shared -- HEAD^..HEAD
... # remove .gitrepo as well
$ git log --format=%T
<tree hash>
$ git commit-tree <tree hash> -p <.gitrepo.parent> -m "Some message"
<commit hash>
$ git reset --hard <commit hash>
```
Now we have a subrepo/shared/branch with our changes connected to the last pull parent. So from here we can do a git merge.

**Comment by ingydotnet** [comment]
@grimmySwe, I'll need to pore over this more, later. One thing I don't quite
get is why you'd ever want to not squash.

As long as you can fetch the upstream commit in question, you always have the
history.

Educate me! :)

**Comment by grimmySwe** [comment]
@ingydotnet 
To clearify this proposal, currently we have this situation:

- Pull
   - Uses merge-base to find equal tree hashes as common ancestor
   - Use rebase to apply changes from the subrepo
   - Creates a squashed commit in your project
   - Updates .gitrepo
- Push
   - You need to be up to date
   - Uses merge-base to find equal tree hashes as common ancestor
   - Use rebase to apply changes into subrepo
   - Commits are left as is, not squashed together
   - Can update .gitrepo with -u

Want I am suggesting is the following:
- Pull (Get latest changes from subrepo)
   - Use the .gitrepo.commit as common ancestor
   - Use merge between subrepo and project
   - Creates a squashed commit in your project (same as now)
   - Update .gitrepo.commit
- Push (put your state in subrepo)
   - You need to be up to date
   - Simply write your latest state to subrepo
   - Creates a squashed commit in the subrepo
   - Update .gitrepo.commit in project

Further extension to this would be to add a --no-squash flag to pull:
- Push --no-squash (put your commits into the subrepo)
   - You need to be up to date
   - Create a correct graph in the subrepo of your commits (so you get real merge commits inside the subrepo) from the .gitrepo.last_push
   - Update .gitrepo.commit and a new .gitrepo.last_push

Even more:
Add a subrepo configuration, so that you can set a squash/no-squash setting in the actual subrepo so that the default behaviour is set for a subrepo.

**Comment by grimmySwe** [comment]
Also consider #144, using filter-branch on an unknown amount of commits might be very slow.

**Comment by grimmySwe** [comment]
Regarding the push case without squashing: If you store the last push in .gitrepo.

First perform a parent-filter subdirectory from the latest push to the current position. Now you can step for each commit, look into which commit it points to in .gitrepo and when that changes add it to the parent list. Finally perform the merge and the push the entire tree into subrepo. That way you would get a complete version in subrepo, containing actual merge commits.
Finally you update the local .gitrepo with new last push and latest commit.

**Comment by grimmySwe** [comment]
@ingydotnet @jrosdahl I have created a first attempt at this (Branch issue/142) and need some feedback.
Note: I haven't bothered about the unittest/documentation yet

Status:
- both push/pull are squashed. 
- You can't push unless you have recently pulled. 
- When you create the subrepo/branch it creates a new node with the .gitrepo.commit as parent.
- Merge is used instead of rebase
- subrepo push will also create a local commit updating the .gitrepo file

Feedback needed:
Currently the squashed push message is simply PLACEHOLDER... What would you consider the best approach:

1. A generic message like the one used in pull/commit
2. User defined with -m/editor opens if -m is missing

I guess the ultimate would be some kind of hint regarding the content, but this is a little bit tricky as you don't really now what the content is right now.

The immediate benefits of this solution:

- You can have a local change and still pull in changes without having to solve the same conflicts over and over again.
- You can work in push mode, performing consecutive pushes as long as no one else pushes.
- "Constant" time operation for the branch creation. You always use the .gitrepo.commit pointer so you don't have to iterate the tree.

Drawbacks:

- Currently push commits are squashed
- Not all commit messages are relevant

**Comment by robe070** [comment]
For our use case we absolutely must not squash push commits. The real commit messages must be retained in the subrepo history. So I would like to see an option for this and that not squashing is the default as it is the current behaviour.

**Comment by grimmySwe** [comment]
@robe070 I am working on the next step of this, where you don't get squashed pushes (unless wanted) and where you hopefully get a logically correct tree in in the subrepo. One of the drawbacks with the current implementation is that when you perform rebases they will end up in a somewhat unlogical order, so the history in the subrepo might be strange. 

Not squashing actually solves another issue that you don't have to generate the push commit messages.

**Comment by grimmySwe** [comment]
Note to self:
subrepo branch --no-squash => create branch with all the changes since last push
subrepo branch --squash => create branch with only the current state

So to find out what you have done since last push you can branch and then diff.

**Comment by ingydotnet** [comment]
@grimmySwe I appreciate this effort. I'm taking the back seat for now, since these things aren't fixing my personal pains.

I wonder if we should try a google hangout with some of the key people, sometime? Figure out a roadmap. I'd be up for it.

**Comment by grimmySwe** [comment]
I actually managed to create a branch command that rebuilds the graph so that when I push you get a better view of what has actually happened in the subrepo.

I bet that there are tons of holes in the implementation but at least I got something to work on. @robe070 If you have some spare time please let me know what you think.  

Status (at least the intention...), consider this a proof of concept:

- subrepo branch will now create a complete graph.
- subrepo branch will use the .gitrepo.parent as starting point
- .gitrepo.parent is only updated on push
- If you want to use squash, we simply override .gitrepo.parent to be HEAD^
- push will actually push the entire graph built by branch command

You will get some additional commit messages (pull/push/commit) inside the subrepo but that is for another day.

**Comment by robe070** [comment]
@grimmySwe As a user with a large repo where subrepo commands take of the order of 5 minutes minimum, and using current functionality that has worked so far, and I have no local changes to the subrepo - all should be in sync -, and I want the push to not squash, what will I gain? 

Can I expect any base performance improvements? With both the `subrepo push' and `subrepo pull`, the bulk of the time is spent in `git reset --hard`.

Some times:
subrepo has a total of about 100 commits
parent repo has about 32,000 commits
First pull after clone, 1 new commit, approx 5 minutes
2nd pull, no new commits, 11s (looks like it quickly works out there is nothing to pull. Cool!)
Every push, no new commits, 2m 33s (shame that push can't work out there is nothing to do earlier too)

**Comment by grimmySwe** [comment]
@robe070 Thanks for your input! It is very good to have other people review what you do.

I work in a large corporation where we currently run svn as for version control (600k commits), but we are in the progress of replacing that with git. So in that process we evaluated the options for replacing svn:externals and we really liked the concept of git-subrepo.

With that said I want to assure you that everything we do is mainly aimed at having git-subrepo interact with large repositories/subrepos in a good way.

The aims of this redesign is:
- Try to make all operations constant time. So as little as possible are dependant on the actual size of the repository. As you say there are operations like reset/checkout that always will be size dependant, but subrepo should at least not add another layer of time.
- Make it possible to choose workflow for push, squash/no-squash. If you have subrepos with constant activity and use feature branches it might be better to actually squash the incoming so that it's easier to track the progress.
- If you use no-squash for push the history in itself should be easier to understand, currently the rebasing can make it a little hard to understand things. If you work on a change in a subrepo and don't want to commit it right away but you still want to pull in new changes to subrepo, when you push the commits you push will replay things a little out of order, starting with a step back and then retracing the steps back until you reach the HEAD again.

**Comment by robe070** [comment]
#158 has a possible issue.
Following on from that, repeating `git subrepo pull` correctly identified the subrepo was up to date and a subsequent `git subrepo push` identified no new commits to push, in a similar time to previous `master` code

**Comment by robe070** [comment]
Issue/142 testing: git subrepo push output after pulling new commits into local tree with no changes relevant to subrepo 
Many fatals similar to the `git subrepo pull` above:
[gist: fatals reported but still completes as expected](https://gist.github.com/robe070/0029b430fa29c8eb5a92)

**Comment by robe070** [comment]
Are there many new commits being created everyone a `git subrepo push` is executed? Is the repo going to be weighed down by orphaned commits?

**Comment by robe070** [comment]
When there are new commits, but no changes, a commit with new .gitrepo is still created.

I also saw a `.gitrepo`-only commit created when I thought there were no commit differences. Sorry, no detail.

Background: in creating git repos from Vault repo, the subrepo is mirrored to two git repos which each have the same subrepo and thus the same commit exists in both. So a very valid scenario, though not a large throughput of changes to this subrepo.

**Comment by grimmySwe** [comment]
First of all, thanks a lot for testing!

> Following on from that, repeating git subrepo pull correctly identified the subrepo was up to date and a subsequent git subrepo push identified no new commits to push, in a similar time to previous master code
> Every single create new commit seems to have failed (and look to be the same shas as in previous run). Does that imply there are no new commits? Can the check for 'something to push' be earlier?

In 9e9067df668622d04a334c380517e716ca061b4b I updated the filter-branch parameters so it should be even faster now.

> Many fatals similar to the git subrepo pull above: gist: fatals reported but still completes as expected Note that second git subrepo push still has fatal errors. If fatals are expected, then maybe a check to validate if the sha is a valid object name may improve performance, and eliminate false negatives.

Could you please retest this with the latest 073b7bef91536107af6ffa17a5eeb446b1542202 as I have done some changes. If it doesn't solve the problem I will need to look into it further, strange though that it reports it as unknown, the commits are taken from a simple ```$(git rev-list --reverse --first-parent "$subrepo_parent"..HEAD)``` 

> Are there many new commits being created everytime a git subrepo push is executed? Is the repo going to be weighed down by orphaned commits?
A push creates a new commit in the parent repo so that the .gitrepo file is updated with correct information. The theory is that when you push next time this push will be silently removed by the filter-branch --prune-empty, but I haven't had time to test it thoroughly yet.

> When there are new commits, but no changes, a commit with new .gitrepo is still created. I would prefer it to detect there are no new changes.
This new design creates a new commit with the parent contents, connects it to the .gitrepo.commit found in the fetch and then performs a merge between the two of those. So the need for new commits are determined of that outcome. For the changes on parent side a filter-branch --prune-empty will remove all commits that only changes .gitrepo file, although I don't know if --prune-empty works well with merge commits yet. On the suprepo side every commit is treated as valid.

**Comment by robe070** [comment]
073b7be does not show the fatals. Cool.

Time has reduced from 5m9.408s with prior test above to 4m26.956s with this commit.

The messages in your new code seem to be missing the git commands. Only verbose messages are displayed:
```
$ time git subrepo push LANSA/Imports/DeveloperPack --verbose --debug
[snip]
* Create subrepo branch 'subrepo-push/LANSA/Imports/DeveloperPack'.
  * Check if the 'subrepo-push/LANSA/Imports/DeveloperPack' branch already exists.
  * Create new commits with parents into the subrepo fetch
  * 871765f140efbd56b6158ba15e010f7db6cb2f14 => 9e4659c817bbefefcc5b1629aa81e453ee12fea4
  * Create a new commit  -p 9e4659c817bbefefcc5b1629aa81e453ee12fea4
  * b76bd4dc09eb985c88d233558851fc9f54308acb => 9e4659c817bbefefcc5b1629aa81e453ee12fea4
  * Create a new commit -p 15b0d38e74106a241f217d143603909b109d5966
  * Create branch 'subrepo-push/LANSA/Imports/DeveloperPack' for this new commit set 135aa18cc2e0b6da8be8b50e53d743d285285dba. 
[snip]
```

**Comment by grimmySwe** [comment]
Strange, when I run the commands I get both.

```
$ git subrepo push bar --verbose --debug
>>> git rev-parse --verify HEAD
* Fetch the upstream: ../bar (master).
  * Fetch '../bar' (master).
>>> git fetch --quiet ../bar master
  * Get the upstream subrepo HEAD commit.
>>> git rev-parse FETCH_HEAD^0
  * Create ref 'refs/subrepo/bar/fetch'.
* Check upstream head against .gitrepo commit.
* Deleting old 'subrepo-push/bar' branch.
>>> git branch -D subrepo-push/bar
* Create subrepo branch 'subrepo-push/bar'.
  * Check if the 'subrepo-push/bar' branch already exists.
  * Create new commits with parents into the subrepo fetch
  * 21238179219ac6aa43cc667a7c7deaca605b9fa2 =>
06983df90d221231bc3aeda3c0a0fd195c972faf
  * Create a new commit  -p 06983df90d221231bc3aeda3c0a0fd195c972faf
  * Create branch 'subrepo-push/bar' for this new commit set
127d6f1abb6e7241a6c393259bce3f4566bf61b5.
>>> git checkout -B subrepo-push/bar
127d6f1abb6e7241a6c393259bce3f4566bf61b5
  * Remove the .gitrepo file from
06983df90d221231bc3aeda3c0a0fd195c972faf..HEAD
>>> git filter-branch -f --prune-empty --tree-filter rm -f .gitrepo --
06983df90d221231bc3aeda3c0a0fd195c972faf..HEAD
  * Create ref 'refs/subrepo/bar/branch'.
>>> git checkout master
* Make sure that 'subrepo-push/bar' exists.
* Check if we have something to push
Subrepo 'bar' has no new commits to push.
```

2016-03-04 21:09 GMT+01:00 Rob Goodridge <notifications@github.com>:

**Comment by grimmySwe** [comment]
Strange, when I run the commands I get both.

```
$ git subrepo push bar --verbose --debug
>> git rev-parse --verify HEAD
* Fetch the upstream: ../bar (master).
  * Fetch '../bar' (master).
>> git fetch --quiet ../bar master
* Get the upstream subrepo HEAD commit.
>> git rev-parse FETCH_HEAD^0
* Create ref 'refs/subrepo/bar/fetch'.
* Check upstream head against .gitrepo commit.
* Deleting old 'subrepo-push/bar' branch.
>> git branch -D subrepo-push/bar
* Create subrepo branch 'subrepo-push/bar'.
  * Check if the 'subrepo-push/bar' branch already exists.
  * Create new commits with parents into the subrepo fetch
  * 21238179219ac6aa43cc667a7c7deaca605b9fa2 =>
06983df90d221231bc3aeda3c0a0fd195c972faf
  * Create a new commit  -p 06983df90d221231bc3aeda3c0a0fd195c972faf
  * Create branch 'subrepo-push/bar' for this new commit set
127d6f1abb6e7241a6c393259bce3f4566bf61b5.
>> git checkout -B subrepo-push/bar
127d6f1abb6e7241a6c393259bce3f4566bf61b5
  * Remove the .gitrepo file from
06983df90d221231bc3aeda3c0a0fd195c972faf..HEAD
>> git filter-branch -f --prune-empty --tree-filter rm -f .gitrepo --
06983df90d221231bc3aeda3c0a0fd195c972faf..HEAD
  * Create ref 'refs/subrepo/bar/branch'.
>> git checkout master
* Make sure that 'subrepo-push/bar' exists.
* Check if we have something to push
Subrepo 'bar' has no new commits to push.
```

**Comment by robe070** [comment]
I'm only talking about the section below. Between these two git commands there aren't any others?
```
>>> git branch -D subrepo-push/bar
* Create subrepo branch 'subrepo-push/bar'.
  * Check if the 'subrepo-push/bar' branch already exists.
  * Create new commits with parents into the subrepo fetch
  * 21238179219ac6aa43cc667a7c7deaca605b9fa2 =>
06983df90d221231bc3aeda3c0a0fd195c972faf
  * Create a new commit  -p 06983df90d221231bc3aeda3c0a0fd195c972faf
  * Create branch 'subrepo-push/bar' for this new commit set
127d6f1abb6e7241a6c393259bce3f4566bf61b5.

>>> git checkout -B subrepo-push/bar
```

**Comment by grimmySwe** [comment]
@ingydotnet I missed to use the RUN method on some places to get better logging on all git commands but I can't figure out how to replace this:
`local output=$(git log -n 1 --format=%B "$commit" | git commit-tree -F - $first_parent $second_parent "$commit":"$subdir")`

Is there a good way? Or will the pipe simple mess things up?

**Comment by grimmySwe** [comment]
More cleanup c26b8cd closing in all tests passing again.

**Comment by robe070** [comment]
@GrimmySwe I'm presuming that I need to wait for the tests to fully pass
before continuing tests? Or do you have some specific test requirement of
me before that? Any ideas when you may be able to achieve all tests passing?

warm regards

Rob Goodridge

**Comment by robe070** [comment]
Using c26b8cd.
The commit message is prefixed with what looks like `--debug` output. Is this intentional?
```
>>> git rev-parse --short 9e4659c817bbefefcc5b1629aa81e453ee12fea4
git subrepo pull DEMODEV
```

**Comment by grimmySwe** [comment]
@robe070 Thanks for your feedback!

> The commit message is prefixed with what looks like --debug output. Is this intentional?

No, I think it was a side effect of me adding additional debug output.

> Any ideas when you may be able to achieve all tests passing?

I think I will have a time slot of me time on tuesday evening CET+1, so my plan is to at least get everything into green at that session.

> I'm presuming that I need to wait for the tests to fully pass
before continuing tests? Or do you have some specific test requirement of
me before that?

I haven't been able to test it a lot myself yet so I would say that save your strength for later this week.

**Comment by grimmySwe** [comment]
Pushed  6fe27bf where all tests go green but...

Current status:

- init then push will currently squash everything into one single subrepo commit
- there will be some pull/commit messages in the subrepo if you merge changes

My todo list right now:

1. init - push should work (not squashing)
2. Add a --squash to push so if you want to you can actually squash (both after init and in general)

Things to think about:

- Commit messages for the auto generated commits, should it be possible to enter more of them manually to get a better history.

Unfortunately the me time is out for today...

**Comment by robe070** [comment]
Thats great!

1. When you have init - push not squashing I'll do some more extensive testing

- When squashing, either pull or push, it would be neater to be able to supply a commit message, and provide a repo config option to make it mandatory to provide that commit message.

**Comment by grimmySwe** [comment]
Pushed a new version of issue/142 branch (ef32f63)

Current status

- All tests go green.
- Init doesn't do very much, simply create the .gitrepo file (Because it doesn't have to or?)
- push after init will use the `git filter-branch --subdirectory-filter`
- Other pushes will regenerate commits by using commit-tree and creating a tree structure with merge commits.
- There is a --squash option for push that will allow you to squash all commits into the subrepo. Note that this can be done for the first push after init as well.

Todo:

1. Work on the commit messages used, make it possible to edit. Consider replacing the commit one with a merge message instead. Will consider this:

> When squashing, either pull or push, it would be neater to be able to supply a commit message, and provide a repo config option to make it mandatory to provide that commit message.

@robe070 Feel free to try out if you have the time.
@ingydotnet What do you think? Even if you don't have any current pains, I haven't found any obvious drawbacks with this "old" method. And I have come across some strange cases with the merge-bsae using tree-hashes.

**Comment by grimmySwe** [comment]
Done some more testing and it seems that I might have found an issue when there are parallel development against one subrepo, problem is that when you generate create your own history from your own data you mangle the commits into new commits, so you can't really determine the same changes that come from the subrepo. So you will actually add duplicate nodes due to this.

I guess this can be solved by always squashing in both directions, in the case where you dont have any other development disturbing you can actually push all commits because then you never need to recreate them again.

Will look into this further.

**Comment by grimmySwe** [comment]
Found that it was only a bug... .gitrepo.parent was not updated on push.
Solved it in (a8ee5a9), now it feels really good!
I still have to work on the commit messages but the tree graph and workflows seems to be ok.

**Comment by ingydotnet** [comment]
@grimmySwe I know you did this before but can you write up the main differences between the old behavior and the new?

I want to figure out if we need to support both the old and the new with config settings.

**Comment by ingydotnet** [comment]
One thing is that I don't like to have any merge commits in my histories. I think many people prefer that as well. Some people, otoh, love merge commits. So I think we at least need a config option for that.

I think we need an option for push to do everything up until the actual push. I always want to know exactly what I'm going to push before I pull the trigger.

Also I just did a test push (using a bogus upstream branch): https://gist.github.com/06d4dcdc843c4722b3d6

Can you tell me why the `git push` in that used `--force`?

**Comment by grimmySwe** [comment]
> Also I just did a test push (using a bogus upstream branch): https://gist.github.com/06d4dcdc843c4722b3d6
> Can you tell me why the git push in that used --force?

I think that is a bug in the code, currently it will always use --force
```local force="$force_wanted" && force=--force```
it should be
```
local force=
"$force_wanted" && force=--force
```

> I think we need an option for push to do everything up until the actual push. I always want to know exactly what I'm going to push before I pull the trigger.

I agree with you there, what do you think of:
```
$ git subrepo branch --fetch bar
# Now subrepo/bar is created with the current suggestion
$ git subrepo push bar subrepo/bar
```

> One thing is that I don't like to have any merge commits in my histories. I think many people prefer that as well. Some people, otoh, love merge commits. So I think we at least need a config option for that.

The main goal of issue/142 has been to make the changes visible and reasonable in time, with the 0.3.0 rebase you get a jumpy history that might be a little hard to track. So yes there will be merge commits included BUT they should only be visible when they are actually needed. So when you pull things that causes merge conflict they will be present due to the fact that something has happened and forced the user to merge two things into one. There is still a `parent-filter --tree-filter --prune-empty` step that should remove any unnecessary commits. Current state is though that the messages are a little hard coded and might not be the best.

About the differences I will try to compile a list soon.
(I think that https://github.com/ingydotnet/git-subrepo/issues/142#issuecomment-182787046 is a good summarize at least.)

**Comment by grimmySwe** [comment]
Ok, lets see:

Both:

- Commits are squashed into parent (pull)
- Commits are preserved into subrepo (push)

0.3.0:

- merge-base based on tree hash to find common ancestors
- rebase/filter-branch to create new structures
- conflicts are solved with rebase
- commits in subrepo is always sequential, no merge commits, note that the content in the commits though are not strictly sequential. Why? Because if you pull several times with a local change, when you push that change you will replay the pull commits as well.

issue/142:

- information in .gitrepo for ancestry (commit and parent)
- commit-tree/filter-branch to create new structures
- conflicts are solved with merge 
- content in subrepo can be parallel, contains merge commits (if needed)
- push will create local commit updating commit/parent
- Commits can be squashed into subrepo (push --squash)

Problems solved with issue/142:

1. You don't have to depend on tree hashes, they become untrustworthy in cases of no changes or reverts.
2. You don't have to iterate the entire history when building new structures
3. You can keep local changes without having to face the same conflicts over and over again
4. If no one else works on the subrepo, you can keep pushing without having to pull
5. You actually see in the subrepo history when things have been done in parallel and you see what merges that has been done. If nothing has been done in parallel, there will be no merge commits.
6. You can decide if you want to squash/no-squash your commits into the subrepo.
7. You can change the .gitrepo file directly if you really need to, allowing you to fix some special cases.
8. If the merge-base tree-hash fails, it's really hard to debug and make it work again. 

Whats left to do in issue/142 (in my opinion):

1. Decide on the commit message structure and content.
   - When you push you might want to make it a custom message, 
   - Merge commits might need to be better and maybe custom
2. Probably clean up the code...
   - I know that I am not a born bash hacker so some of my code can surely be improved/simplified
   - I think I might have mixed the variables a bit refs_* vs commit_*

There might also be an interest in creating some sort of subrepo setting that can be persistent for the actual subrepo. Let's say that you want to force everyone to squash commits into the subrepo (or the opposite, that you want every commit). There might be more things in the future, like a commit message template that you want to use.

Note about merge commits:
The aim with issue/142 is still to keep the subrepo history as clean as possible. If there is only one source that pushes new things there should be no merge commits. These will only appear when thing happen in parallel and there is actual merges done. For me the merge commits signal that there has been merges and the content will show what has been done.

**Comment by ingydotnet** [comment]
@grimmySwe I wrote an email to your commit address. Did you get that? I'll comment more in a bit. I have some concerns here.

**Comment by ingydotnet** [comment]
@grimmySwe My observations are:

* Didn't we used to base pulls off of info in the .gitrepo file? Then we went to merge-base.
  * Are we going back?
  * If so, are we losing the reasons we went to merge-base in the first place? 
  * Are we getting rid of merge-base?
* I don't want to become dependent on merge-only or rebase-only. Does this change do that?
* I like the idea of a --squash for push.

**Comment by grimmySwe** [comment]
> Didn't we used to base pulls off of info in the .gitrepo file? Then we went to merge-base. Are we going back?

Your original 0.2.3 was based on the .gitrepo file, it had some flaws with forgetting changes that I thought could be better handled with the merge-base on tree has. It turned out though that the merge base is quite fragile and will have a performance dependency on the repo size. The information in .gitrepo file is actually enough and often better, because it will show what has actually happened. If you perform a clone/pull/push it will be updated with a reliable information. In the merge-base case you try to recreate the situation by "guessing".

>   If so, are we losing the reasons we went to merge-base in the first place?

I would say that you don't lose anything, instead you actually gain several things:

- You can handle reverts and similar tree hashes. .gitrepo is the one and only interpretation.
- Simplicity, .gitrepo contains the information that your commands will use. You don't have to rely on the merge-base finding a good spot.
- New .gitrepo method should have solved all the previous shortcomings that I know of.
- Flexibility, you can change .gitrepo manually to achieve new things (who knows, there might even be demand to add more subrepo commands to actually change the content in the .gitrepo)

>   Are we getting rid of merge-base?

Yes, I don't see any use of it right now. You can of course leave it as a tool for finding similar tree hashes but personally I think it would be best to skip that. If we don't actually use it it's unnecessary to maintain the code/tests.

> I don't want to become dependent on merge-only or rebase-only. Does this change do that?

My current implementation is merge-only, I don't know if rebase can be used as an alternative. It might be that due to the rebase nature of reimplementing changes will show the previous problems with losing changes.
In the merge case you plant your states into the fetched tree and then use merge to combine them, for rebase you try to reapply changes for a certain range of commits and that will set demands on actually setting the range correctly.

> I like the idea of a --squash for push.

Good :-) it was actually an easy thing to accomplish, simply move the .gitrepo.parent up to previous commit. I guess that the hard part will be generating good commit messages.

**Comment by robe070** [comment]
Its looking good!

A comment on `subrepo init` and `subrepo push`. #144 improved the speed of the combined process. I would like any change in this area to consider #144 and its use case.

**Comment by ingydotnet** [comment]
@grimmySwe I force pushed issue/119 after rebasing it over master.

Regarding update on push by default, I don't want that. I like to try pushing
to other remotes or branches, just for testing.

We can add a config key:

    git config subrepo.auto-update true

that can provide the behavior.

Other possible config options we talked about:

* `subrepo.pull-rebase true`
* `subrepo.push-squash true`

**Comment by grimmySwe** [comment]
> A comment on subrepo init and subrepo push. #144 improved the speed of the combined process. I would like any change in this area to consider #144 and its use case.

I agree with you and I think that it's been even faster now.

- Init, will simply create remote and a local .gitrepo file (no filter branch operation)
- First push, will do the actual work with performing a filter-branch and that should be with visible progress.

**Comment by grimmySwe** [comment]
@ingydotnet I agree that update should be a config option.

What do you think of this:

- Subrepo settings, available in a config file inside the subrepo so that the owner of the subrepo decides the format of the incoming data (not in the .gitrepo file as it's not available in the subrepo)
   - `subrepo.method [merge/rebase]`, Select if you want merges to be visible, also decides how conflicts are handled (by merge or rebase)
   - `subrepo.push-squash [true/false]`, Do you want all commits or squashed
- User preferences
   - `subrepo.push-auto-update true/false`, Do you want the .gitrepo file to be updated when you push

I will look into having both merge and rebase available, but I think that none of them will use the merge-base. As I have created a way to generate the commits I think that it should work to simply use rebase on that instead.

**Comment by ingydotnet** [comment]
@grimmySwe I'm not ready to add another file, and I don't see it as that critical. If you have permissions to push, you should be aware of the current upstream policies. And you could break them with subrepo doing push by hand anyway. This only helps prevent mistakes. Add a new file is too much to ask here. If such a file becomes a common thing outside subrepo, we can consider leveraging it.

**Comment by ingydotnet** [comment]
@grimmySwe I'd like to meet again today. Perhaps at 21:00 your time. I have some ideas and I think I understand where we should go with things.

**Comment by grimmySwe** [comment]
> I'm not ready to add another file, and I don't see it as that critical. If you have permissions to push, you should be aware of the current upstream policies. And you could break them with subrepo doing push by hand anyway. This only helps prevent mistakes. Add a new file is too much to ask here. If such a file becomes a common thing outside subrepo, we can consider leveraging it.

I think you are right here, it's nothing critical and right now it will only make things more complex!

> I'd like to meet again today. Perhaps at 21:00 your time. I have some ideas and I think I understand where we should go with things.

Sounds good, I'll notify you by IRC when I am online.

**Comment by grimmySwe** [comment]
Note to self:
Consider the init command, should it actually make an internal push? Why: because the first push is a little special, it should use `filter-branch subdirectory` to get a good start.

So init => make dir and push to the remote.

Is there a reason to make init without actually doing the first push? You can't use any commands (pull/push/fetch/branch) without a working remote.

**Comment by robe070** [comment]
Its looking good!

A comment on `subrepo init` and `subrepo push`. #144 improved the speed of the combined process. I would like any change in this area to consider #144 and its use case.

**Comment by stsquared99** [comment]
Should the .gitrepo file be auto updated when pushing as well?
After I push changes to a subrepo, the local subrepo is technically up-to-date with the remote, but because the .gitrepo file hasn't been updated yet, when I try to push again, it says "There are new changes upstream, you need to pull first." It seems like one extra step that should be done automatically rather than manually everytime.
Does that make sense?

**Comment by grimmySwe** [comment]
@stsquared99 in the issue/142 branch the .gitrepo file is updated when you push. As you say when you push you become up to date.

@robe070 @stsquared99 
I have created a state diagram:
https://docs.google.com/drawings/d/1gMjINgIEw0lPAAHKPhKMNVjLI2PE9RORGV3BGOTEUHo/edit?usp=sharing
And some example use cases:
https://github.com/ingydotnet/git-subrepo/wiki/issue-142-git-subrepo-use-cases-workflows

In this diagram I try to describe how it works in issue/142 with one change, push should only update .gitrepo when push --update is used. The problem with this is though that it will make the state diagram a little ambiguous, to me I would prefer to have push always update the .gitrepo file.

**Comment by stsquared99** [comment]
I see it now, thanks.

**Comment by grimmySwe** [comment]
@ingydotnet I have tried to implement a rebase variant but I can't get it to work properly, one of my main issues is that with rebase I end up solving the same conflicts twice. First I need to solve them when I pull the data, then I need to solve them again when I push as I can't use the previous information without losing commits (not changes). For the merge case a pull will update the state and I can use that state when I push without having to resolve conflicts.

**Comment by skevy** [comment]
Hey guys -- just wondering what the status was of this rewrite. I've been using the issue/142 branch and it's been working well for my use cases...but just wanted to know if there's a chance this work is gonna make it into master soon?

**Comment by grimmySwe** [comment]
@skevy Been a lot of other stuff lately, I can't really say when it's hitting master. There are some thinking to be done on the commit messages first.

**Comment by ingydotnet** [comment]
@skevy I'd like to get this resolved but it's a big move and I want to make sure I get it right.

Can someone show me a public repo that has used the #142 pull style? All my personal code bases attempt to keep flat (rebased) commit histories. I want to see how this will look and work. I could try to emulate it but it would save time if someone could send me links.

@grimmySwe is it possible to pick one or the other? The way the current style works is to look for a common tree object rather than a common commit object. I realize that you found issues with multiple commits having the same tree SHA, but I can live with that.

Here's an idea that comes to mind...

We add a `--policy=rebase` and a `--policy=merge` flag to `pull` (and thus `push`) and require it if the `.gitrepo` has no entry:

```
        policy = merge
```

then we update the `.gitrepo` file (maybe wanting the `--update` option).

This allows us to keep both behaviors and possibly add more later. I feel most comfortable with this.

We _could_ default to `policy = rebase` and just set the file on `pull`; but I'd rather do that a little bit later.

**Comment by grimmySwe** [comment]
So let's think a little on the commit messages:
Pre-142 behaviour with the rebase:

- Pull creates a single "squashed" commit with a standard message.
   - If no conflicts it will be marked "pull"
   - If there are conflicts it will be a "commit"
- Push let's you rebase your commits into the subrepo, filter branch command will remove empty commits. In most cases this will make the pull commits go away. If changes has happened in the actual pull commits, they will appear in the subrepo.

With 142 there is different approach using merge instead of rebase:

- Pull still creates a single "squashed" commit with a standard message.
   - (A) If no conflicts it will be marked "pull" (There can still been a merge!)
   - (B) If conflicts it will be a "commit"
- Push will always add the new merge nodes, pull/commit to subrepo and a local repo push commit
   - (C) Merge commit into subrepo
   - (D) Push commit into local repo

My opinion:

1. (A) should be Ok being a "pull" when there has been no merge
2. (A/B) should be something else when there has been a merge (Merge/commit)
3. (C) Merge commit in this case will be what (A/B) produces
4. (D) can be a regular push commit, this will never appear in subrepo because it's an empty commit and no merge.

```
Subrepo log from issue/142 branch
$ git log --oneline --graph
* c74a089 Edit bar/gamma                <= push, no merge
* 243f362 Edit gamma
*   ecab0ac git subrepo commit bar   <= push, merge with conflict
|\
| * 81c7504 Edit gamma
* | e681927 Edit bar/gamma
|/
*   469f5eb git subrepo pull bar     <= push, merge but no conflict
|\
| * ae51a95 Edit alpha
* | 77cb9eb Edit bar/gamma
|/
* ce6c413 Add gamma
* b2174fb Add beta
* 694dae8 Add alpha
```

```
Suggestion 
$ git log --oneline --graph
* c74a089 Edit bar/gamma                <= push, no merge
* 243f362 Edit gamma
*   ecab0ac merge, bar subrepo, foo repo   <= push, merge with conflict
|\
| * 81c7504 Edit gamma
* | e681927 Edit bar/gamma
|/
*   469f5eb merge, bar subrepo, foo repo     <= push, merge but no conflict
|\
| * ae51a95 Edit alpha
* | 77cb9eb Edit bar/gamma
|/
* ce6c413 Add gamma
* b2174fb Add beta
* 694dae8 Add alpha
```

A merge commit lives in both repos, so it needs to contain all data.

**Comment by grimmySwe** [comment]
@ingydotnet I thought that a rebase option would be easy using the same base mechanism as the merge in 142, but I can't get it to work in a good way with conflicts. Because when use rebase I end up solving them twice, once when I perform the pull, but then again when I push. Because when I use rebase I can't reuse the structure as can be done in the merge case. You can though argue that conflicts might not be the common case, but then I feel that the merge option is still better as if you don't work simultaneously you will get a linear history.

Hopefully I can find some time soon to test it out properly.

**Comment by jmatsushita** [comment]
Wanted to mention that I've also been using the `issue/142` branch with success. `master` was simply not working for me as I was bumping into various problems. Sorry for no being more precise, but I didn't document my attempts, but I thought that sharing that I have had no problem with `issue/142` would be useful info for others.

Hoping this gets merged!

**Comment by ingydotnet** [comment]
@grimmySwe I have time to consider merging the `issue/142` branch this week. Stay tuned.

**Comment by juddgaddie** [comment]
@grimmySwe I've had a play with issue/142 branch, I like the --squash however for it to be useful I need to have a -m "custom commit message" option rather than the default commit message it uses. Thanks for all your work on this.

**Comment by grimmySwe** [comment]
@juddgaddie I agree with you, there are some improvements needed for the commit message to make it fully operational.

**Comment by grimmySwe** [comment]
Commit messages... been trying to sort out different types and when you want some specifics.

As I see it in git-subrepo you have these commit messages:
1. When you pull and there are no merge needed
2. When you pull and there is a merge (conflict or no conflict)
3. When you push and generate a new push commit to update the local references

Proposal:
For 1 and 3 a pregenerated message should do fine
For 2 I feel that I would like to use the message from the merge resolution process. You can of course add a pregenerated footer.

In addition to this I think that it would be good to add support for the -m/-e flags so that you can perform updates on the commit messages. Basic git allows to pass multiple -m and even add an -e to edit the proposed message.

So git-subrepo could by default generate something like the ones used today, but allow end user to add more stuff with -m and also use -e for an interactive edit. This should be available for clone/pull/commit/push commands. It might be disabled when --all is present.

**Comment by grimmySwe** [comment]
For 2. it might be possible to use the commit -c/-C flags.

**Comment by grimmySwe** [comment]
I performed some tests and thought a little bit more... My conclusions:

1. Enable -m and -e flags for clone/commit/pull.
2. Add a (merge) indicator for pull/commit commit messages so that the auto generated messages indicate that there has happened something extra and that there is a reason for them.

I ignore the -c/-C flags for now, the merges performed inside the subrepo are somewhat artifical and I think that a simple notification will do, if you want more you can always use -e/-m.

So please feel free to test out 5e6ea88 and see what you think.
All tests are running green and there are some added to verify the commit message stuff.

**Comment by grimmySwe** [comment]
@ingydotnet I think have reached "feature complete" on this branch. I haven't updated the documentation yet because I want you opinion on this first.

Current status:
- Use merge (I can't find a rebase variant that works well)
- It's possible to work with push only (if nothing changes in the subrepo)
- It's possible to override commit messages both with -m and -e
- All tests are green
- I am using this in my daily work

**Comment by ingydotnet** [comment]
@grimmySwe good timing. @perlpunk and I took some time to try out this branch. I was hoping that the pull commits would at least be squashed but that seems to not be the case. I think we need to get on a voip call. I would be happy to accommodate you in this regard almost any time this weekend. Can you jump on IRC and let's sync up on a plan?

Cheers and thanks you all your work!

**Comment by ingydotnet** [comment]
I had a review session with @grimmySwe yesterday. #142 looks promising. I updated the doc in 216712d2 and I am going to add support for --rebase and --dry-run for push. After that I think I can move this to master and the world will be a happier place!

**Comment by ingydotnet** [comment]
@grimmySwe I think we can skip today's session. I still have more review work to do. Hopefully we can wrap this up this week. 

Cheers!

**Comment by grimmySwe** [comment]
@ingydotnet Sounds good, let me know if you need an second opinion on your new changes.

Just curious, the --dry-run is that everything except the local .gitrepo update? In that case --dry-run might be a little false as it will actually push thing to other repos. Something like --no-local-update might be better.

**Comment by ingydotnet** [comment]
@grimmySwe: No, `--dry-run` is everything except the push and commit.

**Comment by grimmySwe** [comment]
Added 2e78d5d6 to handle author information, previously it was overwritten with the user that pushed commits to the subrepo. Now author will contain the original author, committer will be the person that actually creates the commits and pushes them.

**Comment by snaddenm** [comment]
I've been having rebase errors while pushing subrepos that looked like they should have pushed without errors. I stumbled across this issue and branch today and when I tried it my problems were gone. Thought I'd drop a comment here as a vote to move ahead and merge this branch into the master branch.

**Comment by mrfelton** [comment]
The work done here looks really promising. It doesn't solve my issue unfortunately, which is that things tend go haywire when you use the `--all` flag.

Here are some of the issues:


> `git-subrepo pull --all --debug --update -b tkp/subrepo-132556355`

When I run this command, there are 2 issues:

1. it's ignoring the `-b` switch and pulling everything form the branch that is listed in the .gitrepo file (develop in my case). The documentation indicates that `-b` should tell it which branch to pul from.

2. The commit messages for each repo that had some changes pulled down (from the wrong branch, as above) are identical. They all reference the same subrepo.

```
commit 1eb19ba7f50f13a4be4e3f1d78f7dea53fd7b089
Date:   Thu Oct 20 10:58:25 2016 +0200

    git subrepo pull (merge) packages/example-api --update --branch=tkp/subrepo-132556355

    subrepo:
      subdir:   "packages/example-api"
      merged:   "658b6f8"
    upstream:
      origin:   "git@github.com:example/example-api.git"
      branch:   "develop"
      commit:   "33de890"
    git-subrepo:
      version:  "0.4.0"
      origin:   "https://github.com/ingydotnet/git-subrepo"
      commit:   "2e78d5d"

commit 4b5cabb57820d903fe729bc308892fd5abe2f65a
Date:   Thu Oct 20 10:58:17 2016 +0200

    git subrepo pull (merge) packages/example-api --update --branch=tkp/subrepo-132556355

    subrepo:
      subdir:   "packages/example-api"
      merged:   "658b6f8"
    upstream:
      origin:   "git@github.com:example/example-api.git"
      branch:   "develop"
      commit:   "33de890"
    git-subrepo:
      version:  "0.4.0"
      origin:   "https://github.com/ingydotnet/git-subrepo"
      commit:   "2e78d5d"
```

```
Toms-MacBook-Pro:example tom$ git whatchanged --pretty --format=short
commit 1eb19ba7f50f13a4be4e3f1d78f7dea53fd7b089

    git subrepo pull (merge) packages/example-api --update --branch=tkp/subrepo-132556355

:100644 100644 63d4d18... ef9327c... M  packages/example-cs/.gitrepo
:100644 100644 1658337... 5bde10e... M  packages/example-cs/TODO.md

commit 4b5cabb57820d903fe729bc308892fd5abe2f65a

    git subrepo pull (merge) packages/example-api --update --branch=tkp/subrepo-132556355

:100644 100644 a8f0514... 14399ab... M  packages/example-api/.gitrepo
:100644 100644 3a5bd43... 3e3e399... M  packages/example-api/src/common/models/person/person.health-check.js
```

As you can see, there are 2 commits, one for each subrepo that was updated, but the commit messages both reference the example-api subrepo (the last commit should be referencing the example-cs repo, since that is what was changed in the commit).

If I run the same command for each subrepo individual without the --all flag then it pulls from the correct branch (the one specified by `-b` and the commit messages are correct (they reference the correct subrepo).

There are similar issues with `push --all` - it ignores the branch flag, pushing everything out to the branch thats listed in .gitrepo.

For now, I'm staying well clear of --all. Everything seems to work as expected without it.

**Comment by mrfelton** [comment]
I thin there is a problem with this PR with is that --update is applied by default when pushing, and there is no way to override this. This makes working across multiple subrepo branches very painful.

**Comment by jmatsushita** [comment]
@mrfelton I also observed the same problem with `--update`. I have also steered clear of `push --all` with this branch. I've been using it for about 6 months now and it would be good to hear what are @grimmySwe and @ingydotnet plans!

**Comment by Darthholi** [comment]
Hello guys!
Subrepo is really a cleverly designed tool, but by trial and error I found out, that it is no use when my repository is big and it takes forever to filterbranch through all the commits.
I suppose that this is one of the redesign questions asked.

So, my question is, what state is the project in, currently?

To note my motivation:
My aim was to compare git-subhistory and git-subrepo and then to use the better one. (Submodule and subtree both obviously lose.)
Now after testing both projects, my aim is to enhance both projects by adding a shortcut for filterbranch. Because subhistory has shorter code, I have already upgraded it (https://github.com/Darthholi/git-subhistory/tree/MC-Devel).
(A dream would be if both projects would 'kind of' merge and take the good things from each other ;) )

**Comment by chrisdtaylor** [comment]
@grimmySwe, I'm still trying to digest  this mega-thread, but hoping maybe you can summarize or at least answer a couple questions.  First, some background on my situation:

I started out using the current 0.3.1 version that is the master branch.  With that version, all subrepo-affecting commits made in a multirepo are rebased onto the subrepo when pushing.  Meaning all commits affecting the subrepo, whether made in the multirepo or in the subrepo, eventually make their way back to the subrepo (no squashing).  For me, having all commits come back to the subrepo intact was a good thing.

More recently I moved to release/0.4.0 after reading that it allowed for merging or rebasing depending on a configuration setting.  That sounded great, but in practice I've encountered some problems for my situation where I want all commits to eventually come back intact to the subrepo.  As mentioned above, the merge approach squashes commits on push from multirepo to subrepo.  And the rebase configuration doesn't seem to actually rebase as far as I can tell but rather does something similar to the merge configuration.

Is there any way to suppress squashing of commits under the merge configuration?

Is the rebase configuration still a work in progress, or do you suspect I'm overlooking something?

**Comment by grimmySwe** [comment]
@chrisdtaylor In 0.4.0 there is no intention to squash commits so if you experience that, please open a problem report so we can look into why it happens.

Rebase operation should be working, BUT it has some draw backs in the long run. We haven't found an easy way to avoid solving conflicts multiple times in the rebase flow. In the merge case we can store the progress in a simpler fashion with a merge commit, in the rebase case there is no good way to bypass doing another rebase when you push making you perform the same conflict resolution again.

**Comment by ingydotnet** [comment]
@grimmySwe I have a possible solution that I'm going to try to work on this week.

The command `git log --graph --decorate --pretty=oneline --abbrev-commit` will produce output like this for a flat history:
```
* 5a89e02 Foo
* c80bfca Bar
* a5d9132 Baz
* ac21bae Quux
```

My idea is that when someone does a complicated merge resolution, we create a merge history that looks like:
```
*   a15698e Merge subrepo commit 'ffd790e'
|\
| * ffd790e Resolved subrepo conflict head commit
* 5ad47b3 Foo
```
We take the head ref of the merge and save it's subrepo parent commit then we orphan that head commit. Later when we pull again we can "unorphan" the commit temporarily using a graft. This should give git the info it needs to not need to redo the conflict resolution.

That's the idea that came to me in a dream last week. I'll let you know if it pans out.

**Comment by grimmySwe** [comment]
@ingydotnet Looks interesting, I don't really get the benefit of creating dead ends with no parents but you probably have a good explanation for that.

I have tried to come up with solutions for the rebase case but it always fall on the fact that when you rebase, you really want to rewrite commits. In the subrepo case you can't go back and change the past for the main repo. If you always push when you pull AND create an extra commit to record this state you can get something similiar to rebase but I guess that you don't always want to push when you pull.

There is although another option that if you don't push when you pull you lose the actual commits (not the changes). So that if you rebase, you need to pull AND push to keep you commits intact in the subrepo. If not you ignore your current commits and squash them into a single last commit.

**Comment by ingydotnet** [comment]
@grimmySwe and all,

I failed to get 0.4.0 out in January. I have a lot of down time next week. Will  try then.

**Comment by grimmySwe** [comment]
This has finally reached master.
